### PR TITLE
Sentient07 new graph2gpu

### PR DESCRIPTION
**Pull request by nouiz**

### CODE REVIEW

File: theano/gpuarray/opt.py [hunk 1]
```diff
@@ -1075,15 +1074,14 @@ def local_gpua_careduce(op, context_name, inputs, outputs):
...
-                    greduce(gpu_reshaped_x))
+                reduce_reshaped_x = greduce(gpu_reshaped_x)
 
                 if reduce_reshaped_x.ndim != outputs[0].ndim:
                     out_shp = []
                     for i in range(x.ndim):
                         if i not in op.axis:
                             out_shp.append(shape_i(x, i))
-                    unreshaped_reduce = reduce_reshaped_x.reshape(
+                    unreshaped_reduce = GpuReshape(len(out_shp))(
```
Comment (Sentient07) [inline theano/gpuarray/opt.py:1084]: @nouiz  This cause tests to fail. I had tested the theano/gpuarray/tests/test_opt.py but didn't give me any failures or errors. But i got these test failures while testing `tests/test_dnn.py`
Shouldn't that line be something like this , 
`unreshaped_reduce = GpuReshape(len(out_shp), None)(*tensor.stack(out_shp))`
I tried that and I get this error, 

```
ERROR (theano.gof.opt): Optimization failure due to: local_gpua_careduce
ERROR (theano.gof.opt): node: Sum{axis=[0], acc_dtype=float64}(HostFromGpu(gpuarray).0)
ERROR (theano.gof.opt): TRACEBACK:
ERROR (theano.gof.opt): Traceback (most recent call last):
  File "/Users/Ramana/projects/theano/Theano/theano/gof/opt.py", line 1820, in process_node
    replacements = lopt.transform(node)
  File "/Users/Ramana/projects/theano/Theano/theano/gpuarray/opt.py", line 196, in local_opt
    new_op = maker(node.op, context_name, node.inputs, node.outputs)
  File "/Users/Ramana/projects/theano/Theano/theano/gpuarray/opt.py", line 1093, in local_gpua_careduce
    *tensor.stack(out_shp))
  File "/Users/Ramana/projects/theano/Theano/theano/gof/op.py", line 602, in __call__
    node = self.make_node(*inputs, **kwargs)
TypeError: make_node() takes exactly 3 arguments (4 given)
```

Comment (Sentient07) [inline theano/gpuarray/opt.py:1084]: @nouiz My bad, I got confused with `tensor.stack`. Your code seemed right. But i'm not understanding the error and why it is happening?

Comment (nouiz) [inline theano/gpuarray/opt.py:1084]: What about:

GpuReshape(len(out_shp))(reduce_reshaped_x, tensor.stack(out_shp))

I forgot to add the reduce_reshaped_x parameter

Comment (Sentient07) [inline theano/gpuarray/opt.py:1084]: One edit, shouldn't that be .. ?
`GpuReshape(tensor.stack(out_shp).ndim)(tensor.stack(out_shp))`

Comment (Sentient07) [inline theano/gpuarray/opt.py:1084]: Oh, i just saw your comment. Didn't see that before commenting

Comment (Sentient07) [inline theano/gpuarray/opt.py:1084]: And yes, what i thought has 1 more dimension than what we need to reshape to.

Comment (nouiz) [inline theano/gpuarray/opt.py:1084]: I don't understand your last comment. It is not the number of dimensions we are talking about here, but the number of argument passed that isn't right.

Comment (Sentient07) [inline theano/gpuarray/opt.py:1084]: Yeah, that i understood. But i thought the error was with the number of dimensions to reshape to.

**Pull request merged**